File: LeetCode/cpp/1684_count_the_number_of_consistent_strings.cc
/*
Difficulty: Easy
*/

#include <array>
#include <string>
#include <unordered_set>
#include <vector>

class Solution {
 public:
  /// Time: O(n), n = # chars in words, Space: O(1)
  int countConsistentStrings(const std::string allowed,
                             const std::vector<std::string>& words) {
    std::array<bool, 26> allowedSet{};
    for (const char ch : allowed) {
      allowedSet[ch - 'a'] = true;
    }

    // Remove a consistent for each word with a letter not in the set.
    int consistents = words.size();
    for (const std::string& word : words) {
      for (const char ch : word) {
        if (!allowedSet[ch - 'a']) {
          consistents--;
          break;
        }
      }
    }

    return consistents;
  }
};

File: LeetCode/rust/1684_count_the_number_of_consistent_strings.rs
/*
Difficulty: Easy
*/

impl Solution {
    /// Time: O(n), n = # chars in words, Space: O(1)
    pub fn count_consistent_strings(allowed: String, words: Vec<String>) -> i32 {
        let mut allowed_set: [bool; 26] = [false; 26];
        for &ch in allowed.as_bytes() {
            allowed_set[(ch - b'a') as usize] = true;
        }

        let mut consistents = words.len() as i32;
        for word in words {
            for &ch in word.as_bytes() {
                if !allowed_set[(ch - b'a') as usize] {
                    consistents -= 1;
                    break;
                }
            }
        }

        consistents
    }
}
